16.0 mig website sale coupon

Honeyxilia opened the pull request:
EDIT May 5th 2025 : The assets (JS, XML, CSS) have been migrated to the new format, and the web UI works once again

fcayre committed:
Import the website_sale_coupon module into the 10.0 branch

fcayre committed:
Add ability to define campaigns without coupon

With this kind of campaign, customers use the campaign name to
subscribe.

This is a hacky implementation, as a coupon is automatically generated
and associated to current sale on customer subscription. This is decent
however as the impact on present and future code should be minimal.

fcayre committed:
Bump version to 2.0.0, add doc and fix web address

fcayre committed:
Fix race condition when displaying the coupon on cart

fcayre committed:
Display coupon code or campaign name on views depending on campaign type

This is done by implementing name_get with this rule for coupons, and
use display_name in python-side views and in the controller for the
javascript view by changing the controller that returns the coupon.

fcayre committed:
Add ergonomic backend sugar (search views, improved tree views) and model indexes used by them

fcayre committed:
Fix Reserved filter

fcayre committed:
[coupon] fr t10n

fcayre committed:
fr t10n improvement

fcayre committed:
Improve fr t10n again

fcayre committed:
Be more resilient to the case of coupon code user input

Coupon code default value is uppercase, so comparing is value with the
uppercased user input is not perfect but a good compromise.

For is_without_coupon campaign, we need to be more thourough at finding
a case-independant match so we do.

fcayre committed:
Bump version to 2.0.1

fcayre committed:
Fix incorrect html markup

fcayre committed:
Make coupon form more visible on cart

fcayre committed:
fr t10n update

fcayre committed:
Bump version to 2.1.0

fcayre committed:
Rename coupon campaign field and inverse its meaning

Old field was coupons_are_exclusive and was turned into its inverse can_cumulate.

fcayre committed:
fr i18n update

fcayre committed:
Implement the possibility to cumulate coupons from the same campaign

This is useful for money-like coupons: Ethikdo, Chèque Vert.

We now use an exception to describe the cumulation error to the final
user (which needs still to be implemented in the front-end).

fcayre committed:
Display coupon cumulation error messages to the front-end user

fcayre committed:
Improve form view display a bit

fcayre committed:
Add help text relative to cumulative coupons reinit

fcayre committed:
First working v12 version

- py3 realted changes
- manifest basic update
- mandatory filter name attribute added
- adapted menu
- removed intermediary sale view
- fixed selector in sale confirmation page

fcayre committed:
Add _description to Campaign and Coupon models

fcayre committed:
Fix useless date to str conversion using unittests

fcayre committed:
Mute expected error logs in tests

fcayre committed:
Update fr t10n of website_sale_coupons

fcayre committed:
Complete review of the portal order page

Remove sidebar and communication on order pages.

Add coupons and sent documents sections.

legalsylvain committed:
[META] enable flake8

legalsylvain committed:
[META] enable black

legalsylvain committed:
[META] enable pyupgrade

legalsylvain committed:
[META] enable isort

fcayre committed:
Display a coupon list in a dedicated tab of sale order form

This list (a tree view), has a button to open the corresponding coupon.
This is specially useful if the coupon view has specific actions, like
in cooperative campaigns, but may be useful for all, so this is part of
the website_sale_coupon module.

fcayre committed:
i18n upadte + fr t10n of the website_sale_coupon module

github-grap-bot committed:
website_sale_coupon 12.0.1.0.1

fcayre committed:
Use the many2many_tags widget with open and create options for coupons on sale order

This make it easier to add a forgotten coupon to a sale.

fcayre committed:
website_sale_coupon 12.0.1.0.2

fcayre committed:
Fix view modification based on a field from a module we do not depend upon

Commit:
Add a wizzard to create multiple coupons at once

fcayre committed:
website_sale_coupon 1.1.0

Commit:
Select coupons once created with wizard

fcayre committed:
website_sale_coupon 1.1.1

fcayre committed:
Add missing i18n pot files

Honeyxilia committed:
[IMP] website_sale_coupon: black, isort, prettier

Honeyxilia committed:
[MIG] website_sale_coupon: Migration to 16.0

Honeyxilia committed:
Added security access configuration for model website_sale_coupon.create_multiple_coupons_wizard

I duplicated the rules present for the default behavior (empty group_id) and the sale_manager group

Honeyxilia committed:
Removed depreciated at_install/post_install decorators

'at_install' and 'post_install' decorators were included in the 'odoo.tests.tagged' decorator, with 'standard' and 'at_install' attributes being enabled by default (can be disabled using '-' in front of the attribute
See https://www.odoo.com/documentation/18.0/developer/reference/backend/testing.html#odoo.tests.tagged

Honeyxilia committed:
Replaced SavepointCase class with TransactionCase class

Honeyxilia committed:
Moved coupon description methods in coupon/sale.order models

As we don't test methods in the controllers sections (since they require a HttpCase which can be tricky to setup), and these methods have no particular reason to be declared in the controller, we moved them to more adequate classes.

Honeyxilia committed:
Improved code coverage

Honeyxilia committed:
Removed unused method

Honeyxilia committed:
Replaced setUp method by setUpClass method

This allows for better performances (only ran once)
I applied the @classmethod decorator on the _create_campaign method too since it's used in the setUpClass method, where the class is passed rather than a specific instance (to my understanding)

Honeyxilia committed:
Migrated old Odoo module to native JS module

Here are notable changes between the old declaration and the new one :
- Javascript files are no longer imported through the view files directly : rather, JS files (and all related assets) need to be declared in __manifest__.py in the asset section (https://www.odoo.com/documentation/16.0/fr/developer/reference/frontend/assets.html )
=> This means for instance that templates declared in assets don't need to be loaded individually
- Components, widgets or services need to be inserted in registries before being used in a given view
- The structure of the file is a "native JS" module, meaning that the file in written in JS without being wrapped in "@odoo.define", before being converted into the Odoo module system by Odoo itself ( https://www.odoo.com/documentation/16.0/fr/developer/reference/frontend/javascript_modules.html )
=> It is still possible to write modules in the old Odoo Module System, but it's recommanded to migrate code to the new format
- RPC calls in Widgets shall use the _rpc method (defined in odoo/addons/web/static/src/legacy/js/core/service_mixins.js ), which can either call a web path, or a model directly)

Honeyxilia committed:
Removed class name from super() call

Honeyxilia committed:
Added widget to initialise coupon lists for coupons-placeholder elements

9184Lucky commented on website_sale_coupon/tests/test_models.py:
```diff
@@ -125,6 +125,13 @@ def test_reserve_and_confirm_coupon(self):
         self.assertNotIn(coupon, so.reserved_coupons())
         self.assertEqual(coupon.used_for_sale_id, so)
 
+    def test_coupon_description(self):
```
Hello, rien de dramatique mais l'ajout des tests aurait pu être fait avec le commit où tu ajoute les méthodes sur les models, ça augmente un poile la lisibilité à mon sens !

Honeyxilia committed:
Added controller safeguards if the cart is empty.

If the cart is empty, then there is no sale order in the request environnement - so, we return an error either in the UI or in the browser console.

fcayre commented on website_sale_coupon/tests/test_models.py:
```diff
@@ -125,6 +132,16 @@ def test_reserve_and_confirm_coupon(self):
+        self.assertIn(coupon, so.reserved_coupons())
+
+        so.action_confirm()  # uses confirm_coupons
+        self.assertNotIn(coupon, so.reserved_coupons())
```
J'aurais bien ajouté aussi un `self.assertIn(coupon, so.used_coupon_ids)` ici, car le `update` de la ligne 50 effectue simultanément les 2 opérations de modification de `used_coupon_ids` et `reserved_for_sale_id` donc les lignes sont "techniquement" couvertes mais pas "logiquement", ce qui reste le plus important (la couverture n'est qu'un indicateur imparfait : si tu fais sauter la ligne d'update sur `used_coupon_ids` la couverture de cette instruction ne change pas mais le code devient buggé !).

Honeyxilia replied:
en effet - je l'ai ajouté

fcayre commented on website_sale_coupon/static/src/js/coupon.esm.js:
```diff
@@ -0,0 +1,95 @@
+                // Done function
+                $button.removeAttr("disabled");
+                $spinner.hide();
+                console.log("use_coupon success! result=%o", result);
```
Ce console.log (présent dans le code initial) me semble doublement incorrect : 
- parce que c'est reserve_coupon qui est appelé non pas use_coupon
- parce que le mot "success" me paraît un peu prématuré (puisqu'on teste la valeur de result.success après) ; on devrait plutôt écrire ici un truc comme : `console.debug("reserve_coupon call returned %s", result)`

fcayre replied:
Je ne suis pas sûr de comprendre la réécriture complète de ce fichier : 
- qu'apporte publicWidget ici ?
- pourquoi l'avoir renommé en .esm.js ? Je n'ai rien vu dans la doc à ce sujet.

Ça rend la relecture compliquée alors qu'il semble que le code soit resté à peu près identique, je me trompe ?

Honeyxilia replied:
1. publicWidget permet de lancer du code lorsque qu'un élément avec le sélecteur `selector` est présent dans la DOM, afin de modifier des attributs
2. Pour le changement d'extention, cela est parce que le module est passé d'un module Odoo (commençant par `odoo.define`) à un module "ES6", ou "Javascript natif" (commençant par `/** @odoo-module */`), ce qui implique un linting différent dans le script de pre-commit, avec ESLint - ainsi, afin de permettre le linting avec ESLint, il faut renommer le fichier avec l'extension `.esm.js` 
cf. `.eslintrc.yml` et https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-15.0

fcayre replied:
1. c'est déjà ce qui était fait dans le code précédent, ce n'est donc pas une motivation pour changer
2. OK !

fcayre replied:
Je ne vois rien au sujet de mon 1er message ici (le console.log).

Honeyxilia replied:
Malheureusement, même en changant le code afin qu'il concorde avec les nouvelles spécifications, et qu'il est bien interprêté par le navigateur, il ne se relance pas après chargement, et ainsi il ne récupère pas les coupons et n'ajoute pas l'action de confirmation de formulaire.
Avec publicWidget et l'identifiant `selector`, on peut enregistrer le widget dans le registre d'Odoo afin qu'il soit lancé systématiquement lorsque l'identifiant est détecté

Quand j'avais recherché dans les autres modules d'Odoo, la modification de DOM se faisait majoritairement via publicWidget, et je n'ai pas vu d'exemple de code lancé sans passer par un publicWidget ou un Component.

fcayre replied:
"il ne se relance pas après chargement" => tu n'as pas compris pourquoi donc, il faut essayer de ne jamais contourner une incompréhension, notamment sans demander à tes camarades un coup de main ;-)
À mon avis il est possible de faire marcher le code initial avec une très très petite modification. Ce que ça permet de faire c'est : 
1. progresser ! c'est comme ça qu'on gagne en efficacité sur le long terme : en "perdant" du temps en comprenant les choses en profondeur
2. avoir un vrai choix sur la transformation en PublicWidget : ça devrait être un choix réfléchi et non pas une contrainte

Ici je pense raisonnable de laisser le code en l'état : 
- car en fait on n'a pas le temps de creuser beaucoup, le départ de Luc est une vraie deadline à tenir
- le code obtenu est joli et lisible
- tu as regardé le code des autres et le tien te semble conforme à celui de la communauté

Mais vraiment j'insiste : on ne se laisse pas diriger seul.e par une incompréhension, on a trop à gagner à discuter entre nous pour chercher une solution collective, et à progresser individuellement en ayant de meilleures connaissances mobilisables sur le long terme.

fcayre replied:
En revanche tu ne dis rien sur le console.log, cf. ma 1ère remarque de cette conversation.

Honeyxilia replied:
Bien reçu pour publicWidget c:

Pour console.log, j'ai répondu à ton autre commentaire sur le sujet, en disant que je les ai changés
-> https://github.com/commown/commown-odoo-addons/pull/381/commits/1347d89d593fed27a799026156cb14b3a2ddd60e

fcayre commented on website_sale_coupon/static/src/xml/coupon_templates.xml:
```diff
@@ -3,20 +3,14 @@
-                            class="close"
-                            data-dismiss="modal"
-                            aria-hidden="true"
-                        >&amp;times;</button>
```
Ne semble pas lié à ce commit, dommage de ne pas l'avoir mis ailleurs comme ça tu aurais pu expliquer que ce bouton en présent en doublon.

Honeyxilia replied:
J'ai déplacé leur suppression dans leur commit dédié

fcayre commented on website_sale_coupon/static/src/xml/coupon_templates.xml:
```diff
@@ -38,20 +32,14 @@
-                            class="close"
-                            data-dismiss="modal"
-                            aria-hidden="true"
-                        >&amp;times;</button>
```
idem

fcayre commented on website_sale_coupon/static/src/js/coupon.esm.js:
```diff
@@ -93,3 +83,23 @@ publicWidget.registry.CouponForm = publicWidget.Widget.extend({
+            route: "/website_sale_coupon/reserved_coupons",
+        })
+            .then(function (result) {
+                console.log("used_coupons success! result=%o", result);
```
même remarque qu'auparavant sur le console.log

Honeyxilia replied:
J'ai renommé les messages de log, et j'ai changé la méthode de log à `debug` c:

fcayre commented on website_sale_coupon/controllers/main.py:
```diff
@@ -30,4 +38,8 @@ def reserve_coupon(self, code):
+        if not so:
+            return {"success": False, "reason": "No sale order."}
+
+        return so._sale_coupons_descr()
```
Pourquoi ne pas renvoyer une liste vide ici ? Il me paraît plus sain de modifier `_sale_coupons_descr` en ce sens (si nécessaire).

Honeyxilia replied:
En effet - j'ai bougé les modifications (pour `reserve_coupon` et `reserved_coupons`) dans le modèle - en conséquence, j'ai ajouté des tests pour vérifier le comportement

Honeyxilia committed:
Remove header Close buttons

The removed buttons were present but invisible in the Odoo 12.0 version - however, in Odoo 16.0, aria-hidden no longer applies the `hidden` attribute on the object, and displays the button.
Since it duplicates the action of the Close button and could risk confusing users, we removed the buttons.

Honeyxilia committed:
Changed log type to debug + reworded log messages

Honeyxilia committed:
Used test util method that was previously unused

fcayre commented:
/ocabot rebase

fcayre commented:
/ocabot merge patch
